devito/interfaces.py

```diff
@@ -440,44 +445,126 @@ class DenseData(TensorData):
     def __init__(self, *args, **kwargs):
         if not self._cached():
             self.name = kwargs.get('name')
-            self.shape = kwargs.get('shape', None)
-            if self.shape is None:
-                dimensions = kwargs.get('dimensions')
-                self.shape = tuple([d.size for d in dimensions])
+            self.grid = kwargs.get('grid', None)
+
+            if self.grid is None:
+                self.shape_domain = kwargs.get('shape', None)
+                self.dtype = kwargs.get('dtype', np.float32)
```

Comment:
nitpicking: move after the ``if``

Reply:
Actually no, since I want the `dtype` to be able to override the `grid.dtype` (one might want to have different dtypes per function), but the defaulting value changes from `int32` to `grid.dtype`.